openedx/features/cms/tests/factories.py

```diff
@@ -0,0 +1,74 @@
+import json
+
+from course_action_state.models import CourseRerunState
+from cms.djangoapps.contentstore.tasks import rerun_course as rerun_course_task
+from cms.djangoapps.contentstore.utils import add_instructor
+from opaque_keys.edx.locator import CourseLocator
+from openedx.features.course_card.helpers import get_related_card_id
+
+from xmodule.course_module import CourseFields
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.exceptions import DuplicateCourseError
+from xmodule.modulestore.tests.factories import XModuleFactory
+from xmodule.modulestore import EdxJSONEncoder
+from lms.djangoapps.courseware.courses import get_course_by_id
+
+
+class CourseRerunFactory(XModuleFactory):
+    """
+    A factory for creating rerun from source course
+    """
+
+    # pylint: disable=unused-argument
+    @classmethod
+    def _create(cls, target_class, **kwargs):
+
+        source_course_id = kwargs.pop('source_course_id', None)
+        run = kwargs.pop('run', None)
+        user = kwargs.pop('user', None)
+
+        if source_course_id is None:
+            raise Exception('Source course must be required')
+
+        if run is None:
+            raise Exception('Run must be required')
+
+        if user is None:
+            raise Exception('User must be required')
+
+        org = kwargs.pop('org', source_course_id.org)
+        course = kwargs.pop('course', source_course_id.course)
+        start = kwargs.pop('start', CourseFields.start.default)
+        display_name = kwargs.pop('display_name', None)
+
+        if not display_name:
+            source_course = get_course_by_id(source_course_id)
+            display_name = source_course.display_name
+
+        rerun_course_id = CourseLocator(
+            org=org, course=course, run=run
+        )
+
+        store = modulestore()
+        # verify org, course and run don't already exist
+        if store.has_course(rerun_course_id, ignore_case=True):
+            raise DuplicateCourseError(source_course_id, rerun_course_id)
+
+        fields = dict()
+
+        fields['start'] = start
+        if display_name:
+            fields['display_name'] = display_name
+
+        # Make sure user has instructor and staff access to the destination course
+        # so the user can see the updated status for that course
+        add_instructor(rerun_course_id, user, user)
+
+        parent_course_key = get_related_card_id(source_course_id)
+        # Mark the action as initiated
+        CourseRerunState.objects.initiated(parent_course_key, rerun_course_id, user, display_name)
+
+        return rerun_course_task.delay(
```

Comment:
Please verify that you get response synchronously with your setting

Reply:
`cms/test` environment has a setting [CELERY_ALWAYS_EAGER](http://docs.celeryproject.org/en/3.1/configuration.html#celery-always-eager) set to `True` which makes all celery task in test environment to synchronous.

Hence, I am removing its decorator from code (unit tests) as `CELERY_ALWAYS_EAGER` is already set in environment